protocol/x/ratelimit/keeper/keeper.go

```diff
@@ -218,6 +221,159 @@ func (k Keeper) SetDenomCapacity(
+//	`capacity_diff = max(baseline, capacity-baseline) * (time_since_last_block / period)`
+//
+// This is basically saying that the capacity returns to the baseline over the course of the `period`.
+// Usually in a linear way, but if the `capacity` is more than twice the `baseline`, then in an exponential way.
+//
+//	`capacity =`
+//	    if `abs(capacity - baseline) < capacity_diff` then `capacity = baseline`
+//	    else if `capacity < baseline` then `capacity += capacity_diff`
+//	    else `capacity -= capacity_diff`
+//
+// On a high level, `capacity` trends towards `baseline` by `capacity_diff` but does not “cross” it.
+func (k Keeper) UpdateCapacityForLimitParams(
+	ctx sdk.Context,
+	limitParams types.LimitParams,
+) {
+	supply := k.bankKeeper.GetSupply(ctx, limitParams.Denom)
+
+	capacityList := k.GetDenomCapacity(ctx, limitParams.Denom).CapacityList
+	if len(capacityList) != len(limitParams.Limiters) {
+		// This violates an invariant. Since this is in the `EndBlocker`, we log an error instead of panicking.
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"denom (%s) capacity list length (%v) != limiters length (%v); skipping capacity update",
+				limitParams.Denom,
+				len(capacityList),
+				len(limitParams.Limiters),
+			),
+		)
+		return
+	}
+
+	// Convert timestamps to milliseconds for algebraic operations.
+	blockTimeMilli := ctx.BlockTime().UnixMilli()
+	prevBlockInfo := k.blockTimeKeeper.GetPreviousBlockInfo(ctx)
+	prevBlockTimeMilli := prevBlockInfo.Timestamp.UnixMilli()
+	timeSinceLastBlockMilli := new(big.Int).Sub(
+		big.NewInt(blockTimeMilli),
+		big.NewInt(prevBlockTimeMilli),
+	)
+	if timeSinceLastBlockMilli.Sign() <= 0 {
+		// This violates an invariant (current block time > prev block time).
+		// Since this is in the `EndBlocker`, we log an error instead of panicking.
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"timeSinceLastBlockMilli (%v) <= 0; skipping capacity update. prevBlockTimeMilli = %v, blockTimeMilli = %v",
+				timeSinceLastBlockMilli,
+				prevBlockTimeMilli,
+				blockTimeMilli,
+			),
+		)
+		return
+	}
+
+	// Declare new capacity list to be populated.
+	newCapacityList := make([]dtypes.SerializableInt, len(capacityList))
+
+	for i, limiter := range limitParams.Limiters {
+		// For each limiter, calculate the current baseline.
+		baseline := k.GetBaseline(ctx, supply.Amount.BigInt(), limiter)
+
+		capacityMinusBaseline := new(big.Int).Sub(
+			capacityList[i].BigInt(), // array access is safe because of the invariant check above
+			baseline,
+		)
+
+		// Calculate left operand: `max(baseline, capacity-baseline)`. This equals `baseline` when `capacity <= 2 * baseline`
+		operandL := new(big.Rat).SetInt(
+			lib.BigMax(
+				baseline,
+				capacityMinusBaseline,
+			),
+		)
+
+		// Calculate right operand: `time_since_last_block / period`
+		periodMilli := new(big.Int).Mul(
+			new(big.Int).SetUint64(uint64(limiter.PeriodSec)),
+			big.NewInt(1000),
```

Comment:
I think either could work. Since we eventually need to cast to `big.Int`, I actually like the current way better, which doesn't give possibility for int overflow. Let me know if you disagree!